main/src/ca/uwaterloo/flix/language/phase/Redundancy.scala

```diff
@@ -801,55 +816,63 @@ object Redundancy extends Phase[TypedAst.Root, TypedAst.Root] {
+     */
+    def of(sym: Symbol.EnumSym, tag: Name.Tag): Used = empty.copy(enumSyms = MultiMap.singleton(sym, tag))
 
     /**
-      * Returns an object where the given enum symbol `sym` and `tag` are marked as used.
-      */
-    def of(sym: Symbol.EnumSym, tag: Name.Tag): Used = empty.copy(enumSyms = MultiMap.empty + (sym, tag))
+     * Returns an object where the given defn symbol `sym` is marked as used.
+     */
+    def of(sym: Symbol.DefnSym): Used = empty.copy(defSyms = MultiMapL.singleton(sym, sym.loc))
 
     /**
-      * Returns an object where the given defn symbol `sym` is marked as used.
-      */
-    def of(sym: Symbol.DefnSym): Used = empty.copy(defSyms = Set(sym))
+     * Returns an object where the given sig symbol `sym` is marked as used.
+     */
+    def of(sym: Symbol.SigSym): Used = empty.copy(sigSyms = MultiMapL.singleton(sym, sym.loc))
 
     /**
-      * Returns an object where the given sig symbol `sym` is marked as used.
-      */
-    def of(sym: Symbol.SigSym): Used = empty.copy(sigSyms = Set(sym))
+     * Returns an object where the given hole symbol `sym` is marked as used.
+     */
+    def of(sym: Symbol.HoleSym): Used = empty.copy(holeSyms = MultiMapL.singleton(sym, sym.loc))
 
     /**
-      * Returns an object where the given hole symbol `sym` is marked as used.
-      */
-    def of(sym: Symbol.HoleSym): Used = empty.copy(holeSyms = Set(sym))
+     * Returns an object where the given variable symbol `sym` is marked as used.
+     */
+    def of(sym: Symbol.VarSym): Used = empty.copy(varSyms = MultiMapL.singleton(sym, sym.loc))
 
     /**
-      * Returns an object where the given variable symbol `sym` is marked as used.
-      */
-    def of(sym: Symbol.VarSym): Used = empty.copy(varSyms = Set(sym))
+     * Returns an object where the given variable symbols `syms` are marked as used.
+     */
+    def of(syms: MultiMapL[Symbol.VarSym, SourceLocation]): Used = empty.copy(varSyms = syms)
 
     /**
-      * Returns an object where the given variable symbols `syms` are marked as used.
-      */
-    def of(syms: Set[Symbol.VarSym]): Used = empty.copy(varSyms = syms)
+     * Returns an object where the given variable symbols `syms` are marked as used.
+     */
+    def of(syms: Set[Symbol.VarSym]): Used = empty.copy(varSyms =
+      syms.foldLeft(MultiMapL.empty[Symbol.VarSym, SourceLocation]) {
+        case (mm, sym) => mm + (sym, sym.loc)
+      })
 
   }
 
   /**
-    * A representation of used symbols.
-    */
+   * A representation of used symbols.
+   */
   private case class Used(enumSyms: MultiMap[Symbol.EnumSym, Name.Tag],
-                          defSyms: Set[Symbol.DefnSym],
-                          sigSyms: Set[Symbol.SigSym],
-                          holeSyms: Set[Symbol.HoleSym],
-                          varSyms: Set[Symbol.VarSym],
+                          defSyms: MultiMapL[Symbol.DefnSym, SourceLocation],
```

Comment:
I am worried about this change because you are modifying the semantics of the entire phase -- and only to satisfy one particular use case. 

I would propose a much less invasive change: We could add an extra field:

```
usesOf: ListMap[VarSym, SourceLocation]
```

Yes this means there is a tiny bit of redundancy, but its worth it because it decouples separate concerns.

Reply:
A better name might be occurrences of